README.adoc
= FFMPEG API

A web service for converting audio/video/image files using FFMPEG.

Based on:

* https://github.com/surebert/docker-ffmpeg-service
* https://github.com/jrottenberg/ffmpeg 
* https://github.com/fluent-ffmpeg/node-fluent-ffmpeg

FFMPEG API is provided as Docker image for easy consumption.

== Endpoints

* `GET /` - API Readme.
* `GET /endpoints` - Service endpoints as JSON.
* `POST /convert/audio/to/mp3` - Convert audio file in request body to mp3. Returns mp3-file.
* `POST /convert/audio/to/wav` - Convert audio file in request body to wav. Returns wav-file.
* `POST /convert/video/to/mp4` - Convert video file in request body to mp4. Returns mp4-file.
* `POST /convert/image/to/jpg` - Convert image file to jpg. Returns jpg-file.
* `POST /video/extract/audio` - Extract audio track from POSTed video file. Returns audio track as 1-channel wav-file.
** Query param: `mono=no` - Returns audio track, all channels.
* `POST /video/extract/images` - Extract images from POSTed video file as PNG. Default FPS is 1. Returns JSON that includes download links to extracted images.
** Query param: `compress=zip|gzip` - Returns extracted images as _zip_ or _tar.gz_ (gzip).
** Query param: `fps=2` - Extract images using specified FPS. 
* `GET /video/extract/download/:filename` - Downloads extracted image file and deletes it from server.
** Query param: `delete=no` - does not delete file.
* `POST /probe` - Probe media file, return JSON metadata.

== Docker image

=== Build your own

* Clone this repository.
* Build Docker image:
** `docker build -t ffmpeg-api .`
* Run image in foreground:
** `docker run -it --rm --name ffmpeg-api -p 3000:3000 ffmpeg-api`
* Run image in background:
** `docker run -d -name ffmpeg-api -p 3000:3000 ffmpeg-api`

=== Use existing

* Run image in foreground:
** `docker run -it --rm --name ffmpeg-api -p 3000:3000 kazhar/ffmpeg-api`
* Run image in background:
** `docker run -d --name ffmpeg-api -p 3000:3000 kazhar/ffmpeg-api`

=== Environment variables

* Default log level is _info_. Set log level using environment variable, _LOG_LEVEL_.
** Set log level to debug:
** `docker run -it --rm -p 3000:3000 -e LOG_LEVEL=debug kazhar/ffmpeg-api`
* Default maximum file size of uploaded files is 512MB. Use environment variable _FILE_SIZE_LIMIT_BYTES_ to change it:
** Set max file size to 1MB:
** `docker run -it --rm -p 3000:3000 -e FILE_SIZE_LIMIT_BYTES=1048576 kazhar/ffmpeg-api`
* All uploaded and converted files are deleted when they've been downloaded. Use environment variable _KEEP_ALL_FILES_ to keep all files inside the container /tmp-directory:
** `docker run -it --rm -p 3000:3000 -e KEEP_ALL_FILES=true kazhar/ffmpeg-api`
* When running on Docker/Kubernetes, port binding can be different than default 3000. Use _EXTERNAL_PORT_ to set up external port in returned URLs in extracted images JSON:
** `docker run -it --rm -p 3001:3000 -e EXTERNAL_PORT=3001 kazhar/ffmpeg-api`


== Usage

Input file to FFMPEG API can be anything that ffmpeg supports. See https://www.ffmpeg.org/general.html#Supported-File-Formats_002c-Codecs-or-Features[ffmpeg docs for supported formats].

=== Convert

Convert audio/video/image files using the API.

* `curl -F "file=@input.wav" 127.0.0.1:3000/convert/audio/to/mp3  > output.mp3`
* `curl -F "file=@input.m4a" 127.0.0.1:3000/convert/audio/to/wav  > output.wav`
* `curl -F "file=@input.mov" 127.0.0.1:3000/convert/video/to/mp4  > output.mp4`
* `curl -F "file=@input.mp4" 127.0.0.1:3000/convert/videp/to/mp4  > output.mp4`
* `curl -F "file=@input.tiff" 127.0.0.1:3000/convert/image/to/jpg  > output.jpg`
* `curl -F "file=@input.png" 127.0.0.1:3000/convert/image/to/jpg  > output.jpg`

=== Extract images

Extract images from video using the API.

* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/images`
** Returns JSON that lists image download URLs for each extracted image.
** Default FPS is 1.
** Images are in PNG-format.
** See sample: link:./samples/extracted_images.json[extracted_images.json].
* `curl 127.0.0.1:3000/video/extract/download/ba0f565c-0001.png`
** Downloads exracted image and deletes it from server.
* `curl 127.0.0.1:3000/video/extract/download/ba0f565c-0001.png?delete=no`
** Downloads exracted image but does not deletes it from server.
* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/images?compress=zip > images.zip`
** Returns ZIP package of all extracted images.
* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/images?compress=gzip > images.tar.gz`
** Returns GZIP (tar.gz) package of all extracted images.
* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/images?fps=0.5`
** Sets FPS to extract images. FPS=0.5 is every two seconds, FPS=4 is four images per seconds, etc.

=== Extract audio

Extract audio track from video using the API.

* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/audio`
** Returns 1-channel WAV-file of video's audio track.
* `curl -F "file=@input.mov" 127.0.0.1:3000/video/extract/audio?mono=no`
** Returns WAV-file of video's audio track, with all the channels as in input video.

=== Probe

Probe audio/video/image files using the API.

* `curl -F "file=@input.mov" 127.0.0.1:3000/probe`
** Returns JSON metadata of media file.
** The same JSON metadata as in ffprobe command: `ffprobe -of json -show_streams -show_format input.mov`.
** See sample of MOV-video metadata: link:./samples/probe_metadata.json[probe_metadata.json].


== Background

Originally developed by https://github.com/surebert[Paul Visco].                  

Changes include new functionality, updated Node.js version, Docker image based on Alpine, logging and other major refactoring.


CHANGES.adoc
=== 0.2

* Added extract images and audio endpoints.
* Added probe endpoint.
* Major refactoring.

=== 0.1

* Initial version.

src/routes/test.js
var express = require('express')


var router = express.Router()
const logger = require('../utils/logger.js')

//route to handle file upload in all POST requests


// convert audio or video or image to mp3 or mp4 or jpg
router.post("/",function (req, res,next) {
    logger.debug("path: " + req.path);
    logger.debug("req.params: ");
    for (const key in req.params) {
        logger.debug(`${key}: ${req.params[key]}`);
      }
    logger.debug("res.locals.savedFile: " + res.locals.savedFile);
    res.status(200).send("Test OK.");
          
});

module.exports = router;

src/app.js
const express = require('express');
const app = express();
const compression = require('compression');
const all_routes = require('express-list-endpoints');

const logger = require('./utils/logger.js');
const constants = require('./constants.js');

fileSizeLimit = constants.fileSizeLimit;
timeout = 3600000;

// catch SIGINT and SIGTERM and exit
// Using a single function to handle multiple signals
function handle(signal) {
    logger.info(`Received ${signal}. Exiting...`);
    process.exit(1)
  }  
//SIGINT is typically CTRL-C
process.on('SIGINT', handle);
//SIGTERM is sent to terminate process, for example docker stop sends SIGTERM
process.on('SIGTERM', handle);

app.use(compression());

//routes to handle file upload for all POST methods
var upload = require('./routes/uploadfile.js');
app.use(upload);

//test route for development
var test = require('./routes/test.js');
app.use('/test', test);

//routes to convert audio/video/image files to mp3/mp4/jpg
var convert = require('./routes/convert.js');
app.use('/convert', convert);

var extract = require('./routes/extract.js');
app.use('/video/extract', extract);


require('express-readme')(app, {
    filename: 'index.md',
    routes: ['/'],
});


const server = app.listen(constants.serverPort, function() {
    let host = server.address().address;
    let port = server.address().port;
    logger.info('listening http://'+host+':'+port)
});

server.on('connection', function(socket) {
    logger.debug(`new connection, timeout: ${timeout}`);
    socket.setTimeout(timeout);
    socket.server.timeout = timeout;
    server.keepAliveTimeout = timeout;
});

app.get('/endpoints', function(req, res) {
    let code = 200;
    res.writeHead(code, {'content-type' : 'text/plain'});
    res.end("Endpoints:\n\n"+JSON.stringify(all_routes(app),null,2)+'\n');
});

app.use(function(req, res, next) {
  res.status(404).send({error: 'route not found'});
});


//custom error handler to return text/plain and message only
app.use(function(err, req, res, next){
    let code = err.statusCode || 500;
    let message = err.message;
    res.writeHead(code, {'content-type' : 'text/plain'});
    res.end(`${err.message}\n`);
    
});


logger.debug(JSON.stringify(all_routes(app)));


src/routes/convert.js
var express = require('express')
const fs = require('fs');
const ffmpeg = require('fluent-ffmpeg');

const constants = require('../constants.js');

var router = express.Router()
const logger = require('../utils/logger.js')


//routes for /convert
//adds conversion type and format to res.locals. to be used in final post function
router.post('/audio/to/mp3', function (req, res,next) {

    res.locals.conversion="audio"
    res.locals.format="mp3"
    return convert(req,res,next);
});

router.post('/video/to/mp4', function (req, res,next) {

    res.locals.conversion="video"
    res.locals.format="mp4"
    return convert(req,res,next);
});

router.post('/image/to/jpg', function (req, res,next) {

    res.locals.conversion="image"
    res.locals.format="jpg"
    return convert(req,res,next);
});

// convert audio or video or image to mp3 or mp4 or jpg
function convert(req,res,next) {
    let format = res.locals.format;
    let conversion = res.locals.conversion;
    logger.debug(`path: ${req.path}, conversion: ${conversion}, format: ${format}`);
    if (conversion == undefined || format == undefined)
    {
        res.status(400).send("Invalid convert URL. Use one of: /convert/image/to/jpg, /convert/audio/to/mp3 or /convert/video/to/mp4.\n");
        return;
    }

    let ffmpegParams ={
        extension: format
    };
    if (conversion == "image")
    {
        ffmpegParams.outputOptions= ['-pix_fmt yuv422p']
    }
    if (conversion == "audio")
    {
        ffmpegParams.outputOptions=['-codec:a libmp3lame' ]
    }
    if (conversion == "video")
    {
        ffmpegParams.outputOptions=[
            '-codec:v libx264',
            '-profile:v high',
            '-r 15',
            '-crf 23',
            '-preset ultrafast',
            '-b:v 500k',
            '-maxrate 500k',
            '-bufsize 1000k',
            '-vf scale=-2:640',
            '-threads 8',
            '-codec:a libfdk_aac',
            '-b:a 128k',
        ];
    }

    let savedFile = res.locals.savedFile;
    let outputFile = savedFile + '-output.' + ffmpegParams.extension;
    logger.debug(`begin conversion from ${savedFile} to ${outputFile}`)

    //ffmpeg processing... converting file...
    let ffmpegConvertCommand = ffmpeg(savedFile);
    ffmpegConvertCommand
            .renice(constants.defaultFFMPEGProcessPriority)
            .outputOptions(ffmpegParams.outputOptions)
            .on('error', function(err) {
                logger.error(`${err}`);
                utils.deleteFile(savedFile);
                res.writeHead(500, {'Connection': 'close'});
                res.end(JSON.stringify({error: `${err}`}));
            })
            .on('end', function() {
                utils.deleteFile(savedFile);
                logger.debug(`starting download to client ${savedFile}`);

                res.download(outputFile, null, function(err) {
                    if (err) {
                        logger.error(`download ${err}`);
                    }
                    else {
                        utils.deleteFile(`${outputFile}`);
                    }
                });
            })
            .save(outputFile);
        
}

module.exports = router

src/routes/uploadfile.js
var express = require('express')
const fs = require('fs');
const Busboy = require('busboy');
const uniqueFilename = require('unique-filename');

var router = express.Router()
const logger = require('../utils/logger.js')

//route to handle file upload in all POST requests
//file is saved to res.locals.savedFile and can be used in subsequent routes.
router.use(function (req, res,next) {
    
    if(req.method == "POST")
    {
        logger.debug(`${__filename} path: ${req.path}`);

        let bytes = 0;
        let hitLimit = false;
        let fileName = '';
        var savedFile = uniqueFilename('/tmp/');
        let busboy = new Busboy({
            headers: req.headers,
            limits: {
                fields: 0, //no non-files allowed
                files: 1,
                fileSize: fileSizeLimit,
        }});
        busboy.on('filesLimit', function() {
            logger.error(`upload file size limit hit. max file size ${fileSizeLimit} bytes.`)
        });
        busboy.on('fieldsLimit', function() {
            let msg="Non-file field detected. Only files can be POSTed.";
            logger.error(msg);
            let err = new Error(msg);
            err.statusCode = 400;
            next(err);
        });

        busboy.on('file', function(
            fieldname,
            file,
            filename,
            encoding,
            mimetype
        ) {
            file.on('limit', function(file) {
                hitLimit = true;
                let msg = `${filename} exceeds max size limit. max file size ${fileSizeLimit} bytes.`
                logger.error(msg);
                res.writeHead(500, {'Connection': 'close'});
                res.end(JSON.stringify({error: msg}));
            });
            let log = {
                file: filename,
                encoding: encoding,
                mimetype: mimetype,
            };
            logger.debug(`file:${log.file}, encoding: ${log.encoding}, mimetype: ${log.mimetype}`);
            file.on('data', function(data) {
                bytes += data.length;
            });
            file.on('end', function(data) {
                log.bytes = bytes;
                logger.debug(`file: ${log.file}, encoding: ${log.encoding}, mimetype: ${log.mimetype}, bytes: ${log.bytes}`);
            });

            fileName = filename;
            savedFile = savedFile + "-" + fileName;
            logger.debug(`uploading ${fileName}`)
            let written = file.pipe(fs.createWriteStream(savedFile));
            if (written) {
                logger.debug(`${fileName} saved, path: ${savedFile}`)
            }
        });
        busboy.on('finish', function() {
            if (hitLimit) {
                fs.unlinkSync(savedFile);
                return;
            }
            logger.debug(`upload complete. file: ${fileName}`)
            res.locals.savedFile = savedFile;
            next();
        });
        return req.pipe(busboy);
    }
    next();
});

module.exports = router;

src/routes/extract.js
var express = require('express')
const fs = require('fs');
const ffmpeg = require('fluent-ffmpeg');
const uniqueFilename = require('unique-filename');
var archiver = require('archiver');

const constants = require('../constants.js');
const logger = require('../utils/logger.js');
const utils = require('../utils/utils.js');

var router = express.Router();


//routes for /video/extract
//extracts audio from video
//extracts images from vide
router.post('/audio', function (req, res,next) {

    res.locals.extract="audio"
    return extract(req,res,next);
});

router.post('/images', function (req, res,next) {

    res.locals.extract="images"
    return extract(req,res,next);
});

router.get('/download/:filename', function (req, res,next) {
    //download extracted image
    let filename = req.params.filename;
    let file = `/tmp/${filename}`
    return utils.downloadFile(file,null,req,res,next);
});

// extract audio or images from video
function extract(req,res,next) {
    let extract = res.locals.extract;
    logger.debug(`extract ${extract}`);
    
    let fps = req.query.fps || 1;
    //compress = zip or gzip
    let compress = req.query.compress || "none";
    let ffmpegParams ={};
    var format = "png";
    if (extract === "images"){
        format = "png"
        ffmpegParams.outputOptions=[
            `-vf fps=${fps}`
        ];    
    }
    if (extract === "audio"){
        format = "wav"
        ffmpegParams.outputOptions=[
            '-vn',
            `-f ${format}` 
        ];    
        let monoAudio = req.query.mono || "yes";
        if (monoAudio === "yes" || monoAudio === "true")
        {
            logger.debug("extracting audio, 1 channel only")
            ffmpegParams.outputOptions.push('-ac 1')
        }
        else{
            logger.debug("extracting audio, all channels")
        }
    }

    ffmpegParams.extension = format;

    let savedFile = res.locals.savedFile;

    var outputFile = uniqueFilename('/tmp/') ;
    logger.debug(`outputFile ${outputFile}`);
    var uniqueFileNamePrefix = outputFile.replace("/tmp/","");
    logger.debug(`uniqueFileNamePrefix ${uniqueFileNamePrefix}`);

    //ffmpeg processing...
    var ffmpegCommand = ffmpeg(savedFile);
    ffmpegCommand = ffmpegCommand
            .renice(constants.defaultFFMPEGProcessPriority)
            .outputOptions(ffmpegParams.outputOptions)
            .on('error', function(err) {
                logger.error(`${err}`);
                utils.deleteFile(savedFile);
                res.writeHead(500, {'Connection': 'close'});
                res.end(JSON.stringify({error: `${err}`}));
            })

    //extract audio track from video as wav
    if (extract === "audio"){
        let wavFile = `${outputFile}.${format}`;
        ffmpegCommand
            .on('end', function() {
                logger.debug(`ffmpeg process ended`);

                utils.deleteFile(savedFile)
                return utils.downloadFile(wavFile,null,req,res,next);
            })
          .save(wavFile);
        
        }

    //extract png images from video
    if (extract === "images"){
        ffmpegCommand
            .output(`${outputFile}-%04d.png`)
            .on('end', function() {
                logger.debug(`ffmpeg process ended`);

                utils.deleteFile(savedFile)

                //read extracted files
                var files = fs.readdirSync('/tmp/').filter(fn => fn.startsWith(uniqueFileNamePrefix));
                
                if (compress === "zip" || compress === "gzip")
                {
                    //do zip or tar&gzip of all images and download file
                    var archive = null;
                    var extension = "";
                    if (compress === "gzip") {
                        archive = archiver('tar', {
                            gzip: true,
                            zlib: { level: 9 } // Sets the compression level.
                        });
                        extension = "tar.gz";
                    }
                    else {
                        archive = archiver('zip', {
                            zlib: { level: 9 } // Sets the compression level.
                        });
                        extension = "zip";
                    }

                    let compressFileName = `${uniqueFileNamePrefix}.${extension}`
                    let compressFilePath = `/tmp/${compressFileName}`
                    logger.debug(`starting ${compress} process ${compressFilePath}`);
                    var compressFile = fs.createWriteStream(compressFilePath);

                    archive.on('error', function(err) {
                      return next(err);
                    });
                    
                    // pipe archive data to the output file
                    archive.pipe(compressFile);
                    
                    // add files to archive
                    for (var i=0; i < files.length; i++) {
                        var file = `/tmp/${files[i]}`;
                        archive.file(file, {name: files[i]});
                    }
                    
                    // listen for all archive data to be written
                    // 'close' event is fired only when a file descriptor is involved
                    compressFile.on('close', function() {
                        logger.debug(`${compressFileName}: ${archive.pointer()} total bytes`);
                        logger.debug('archiver has been finalized and the output file descriptor has closed.');

                        // delete all images
                        for (var i=0; i < files.length; i++) {
                            var file = `/tmp/${files[i]}`;
                            utils.deleteFile(file);
                        }

                        //return compressed file
                        return utils.downloadFile(compressFilePath,compressFileName,req,res,next);

                    });
                    // Wait for streams to complete
                    archive.finalize();

                }
                else
                {
                    //return JSON list of extracted images

                    logger.debug(`output files in /tmp`);
                    var responseJson = {};
                    let externalPort = constants.externalPort || constants.serverPort;
                    responseJson["totalfiles"] = files.length;
                    responseJson["description"] = `Extracted image files and URLs to download them. By default, downloading image also deletes the image from server. Note that port ${externalPort} in the URL may not be the same as the real port, especially if server is running on Docker/Kubernetes.`;
                    var filesArray=[];
                    for (var i=0; i < files.length; i++) {
                        var file = files[i];             
                        logger.debug("file: " + file);
                        var fileJson={};
                        fileJson["name"] = file;
                        fileJson[`url`] = `${req.protocol}://${req.hostname}:${externalPort}${req.baseUrl}/download/${file}`;
                        filesArray.push(fileJson);                    
                    }             
                    responseJson["files"] = filesArray;
                    res.status(200).send(responseJson);

                }
            })
            .run();

    }

}

module.exports = router